Synchronize default config values and fix xml generator

mmedenjak opened the pull request:
Here we move the old config compatibility checker to the test code since it is no longer used in production. We also expand on this compatibility checker to include as many configuration as possible and use it to check if the default values in the XML config, the Java config and hazelcast-default.xml are the same. The XML generator tests were already using this compatibility checker so expanding it uncovered some missing config values. The changes :

- fix the "default" value in the hazelcast-config.xsd for the networkConfig/joinConfig/multicastConfig/loopbackModeEnabled from true to false
- fix documentation for the some default values
- add fluent API for generating XML
- add missing tags to xml generator

devOpsHazelcast commented:
Test PASSed.

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java:
```diff
@@ -0,0 +1,852 @@
+        checkHotRestartConfigCompatible(c1, c2);
+
+        checkNetworkConfigCompatible(c1, c2);
+        checkCompatibleConfigs("map", c1, c2, c1.getMapConfigs(), c2.getMapConfigs(), new MapConfigChecker());
```
Nice & arduous job. Would it be possible to homogenize all config checks to be invoked with `checkCompatibleConfigs`? (as far as I could tell, all compatibility checks are done by `ConfigChecker`s)

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/Config.java:
```diff
@@ -1286,7 +1282,7 @@ public Config setDistributedClassloadingConfig(DistributedClassloadingConfig dis
     }
 
-    private <T> T lookupByPattern(Map<String, T> configPatterns, String itemName) {
+    <T> T lookupByPattern(Map<String, T> configPatterns, String itemName) {
```
It's used in the `com.hazelcast.config.ConfigCompatibilityChecker#checkCompatibleConfigs`. Perhaps we can do this differently, haven't thought about it.

mmedenjak commented on hazelcast-spring/src/main/resources/hazelcast-spring-3.8.xsd:
```diff
@@ -222,7 +222,7 @@
+                                            because no items are getting retired. The default value is 10000.
                                         </xs:documentation>
                                     </xs:annotation>
                                 </xs:attribute>
```
I would. Can you suggest something, I am not sure if I can write something meaningful. Something like : 
> Custom classes implementing com.hazelcast.nio.serialization.DataSerializableFactory to be registered. These can be used to speed up serialization/deserialization of objects.

mmedenjak commented on hazelcast/src/main/resources/hazelcast-default.xml:
```diff
@@ -221,7 +221,7 @@
     <jobtracker name="default">
         <max-thread-size>0</max-thread-size>
-        <!-- Queue size 0 means number of partitions * 2 -->
+        <!-- Queue size 0 means the number of processors available to the Java virtual machine -->
```
Ah, I mistakenly thought the comment applied to `max-thread-size`. Will fix.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6723/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6724/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

mmedenjak commented:
@vbekiaris added more missing configs, refactored the xml builder to use a fluent xml generator and addressed the comments. Will rebase after there are no more needed changes.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6861/

devOpsHazelcast commented:
Test PASSed.

jerrinot commented:
@vbekiaris, @bilalyasar, @hasancelik: guys, are you happy with this PR now?

mmedenjak commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.
